File: merge_sort.py
def merge_sort(arr):
    """Sorts an array in ascending order using the merge sort algorithm."""
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left_half = merge_sort(arr[:mid])
    right_half = merge_sort(arr[mid:])

    return merge(left_half, right_half)

def merge(left, right):
    """Merges two sorted sublists into a single sorted list."""
    result = []
    i = j = 0

    # Merge process for combining two halves
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    # Append remaining elements
    result.extend(left[i:])
    result.extend(right[j:])

    return result

# For testing the function 
if __name__ == "__main__":
    test_list = [38, 27, 43, 3, 9, 82, 10]
    print("Sorted array:", merge_sort(test_list))

File: path/to/merge_sort.js
function mergeSort(arr) {
    // Base case: array is already sorted or with a single element
    if (arr.length <= 1) {
        return arr;
    }

    // Split the array into two halves
    const mid = Math.floor(arr.length / 2);
    const leftHalf = arr.slice(0, mid);
    const rightHalf = arr.slice(mid);

    // Recursively sort both halves and merge them
    return merge(mergeSort(leftHalf), mergeSort(rightHalf));
}

function merge(left, right) {
    const result = [];
    let i = 0;
    let j = 0;

    // Merge both sorted halves into a result array
    while (i < left.length && j < right.length) {
        if (left[i] < right[j]) {
            result.push(left[i]);
            i++;
        } else {
            result.push(right[j]);
            j++;
        }
    }

    // Concatenate any remaining elements
    return result.concat(left.slice(i)).concat(right.slice(j));
}

// Example usage
const testArray = [38, 27, 43, 3, 9, 82, 10];
console.log("Sorted array:", mergeSort(testArray));
